### PR TITLE
Attempt to avoid exporting mono univs from Qed proofs

### PR DESCRIPTION
**Pull request by SkySkimmer**
and warn when we must export a constraint between non private univs

Full warning output from compiling stdlib:
https://gist.github.com/SkySkimmer/4bd60e230bb3b3d1f6a6f92bd13ed94f

Code is a bit bugged (especially Derive and Next Obligation don't like
close_proof with keep_body_ucst_separate AFAICT, but without it the
higher layers merge the body and type univ declarations) but gives an
idea of what happens.

Overlays:
- https://github.com/LPCIC/coq-elpi/pull/625

### CODE REVIEW

File: theories/Init/Logic.v [hunk 1]
```diff
@@ -411,7 +411,10 @@ Section Logic_lemmas.
   Qed.
 
   Section equality.
-    Variables A B : Type.
+    (* XXX maybe we should have separate univs for A and B *)
+    Universe equality_u.
```
Comment (SkySkimmer) [inline theories/Init/Logic.v:415]: just trying to make the warnings more readable here

File: theories/Init/Datatypes.v [hunk 2]
```diff
@@ -234,7 +234,9 @@ Register pair as core.prod.intro.
 Register prod_rect as core.prod.rect.
 
 Section projections.
-  Context {A : Type} {B : Type}.
+  (* NB don't seem to need separate universes from the base prod univs
+     (template poly typing rule would force [univ of A <= prod.u0] etc regardless) *)
+  Context {A : Type@{prod.u0}} {B : Type@{prod.u1}}.
```
Comment (SkySkimmer) [inline theories/Init/Datatypes.v:239]: without this we get a couple more constraints in datatypes but not really important

File: theories/Init/Logic.v [hunk 3]
```diff
@@ -547,8 +550,9 @@ End EqNotations.
 Import EqNotations.
 
 Section equality_dep.
-  Variable A : Type.
-  Variable B : A -> Type.
+  Universes f_equal_dep_u0 f_equal_dep_u1.
+  Variable A : Type@{f_equal_dep_u0}.
+  Variable B : A -> Type@{f_equal_dep_u0}.
```
Comment (silene) [inline theories/Init/Logic.v:555]:
```suggestion
  Variable B : A -> Type@{f_equal_dep_u1}.
```
Either that or `f_equal_dep_u1` should be removed.

**Comment by silene** [comment]
It seems like a non-negligible number of opaque universe constraints comes the use of `rewrite`. Perhaps forcing the universe of `eq_ind_r` to be identical to the one of `eq` would improve things. Indeed, the failing lemmas presumably have equalities as hypotheses, so the corresponding constraint is already public and no private constraint would be needed anymore.

**Comment by SkySkimmer** [comment]
>Perhaps forcing the universe of eq_ind_r to be identical to the one of eq would improve things

See new commit and revision 2 of https://gist.github.com/SkySkimmer/4bd60e230bb3b3d1f6a6f92bd13ed94f
tldr about -300 lines (warning location and message take a line each per Qed so probably more like -100 constraints)
(diff isn't only deletions because of building with parallel jobs)

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
I thought it would have produced a slowdown (I don't trust the perf of constraints_for / restrict_universe_context) but apparently it's a bit faster if anything.

**Comment by SkySkimmer** [comment]
NB: Previous rebase was broken in some way

**Comment by ppedrot** [comment]
We should still push this idea further IMHO.
